Update Spack's unparser from cpython, fix package hash instabilities

Jordan474 opened the pull request:
Fix package hash instabilities between python 3.7 and 3.8 https://github.com/spack/spack/issues/45408.

There were 23 builtin packages affected by this. It mostly was AST changes around strings and f-strings. This is important to me because my "fresh" CI/CD is running python 3.11, but my users are running 3.6+.

Package hash for `armcomputelibrary` and `mfem` changed, [see comment](https://github.com/spack/spack/pull/47217#issuecomment-2493473928).

Also, I kind-of went down the rabbit hole of updating Spack's entire unparser.py to mirror `class Unparser` of [`cpython/Lib/ast.py`](https://github.com/python/cpython/blob/main/Lib/ast.py). One can now more easily compare spack's unparser directly to class Unparse of [cpython Lib/ast.py](https://github.com/python/cpython/blob/main/Lib/ast.py): 

```
git clone https://github.com/python/cpython
cp cpython/Lib/ast.py .../spack/lib/spack/spack/util/unparse/cpython-ast.py

cd .../spack/lib/spack/spack/util/unparse
spack style -t black --fix ./cpython-ast.py

your-favorite-diff-tool cpython-ast.py unparser.py  # Skip to the `class Unparse`
```

---

1. Remove differences by backporting one by one cpython ast.py unparser changes to spack's unparser, without breaking python 3.6+ compat, and **no change in current builtin packages hash** with python 3.6 to 3.12:
    - [X] Update style (naming, function order, etc.) to mirror cpython's
    - [X] Update minor code changes
    - [X] Update more sensible changes
    - [X] Drop more Python 2 code and ast nodes
    - [X] Don't change package hash
3. Fix #45408
    - [x] Update major parts of the unparser (might include more that just this fix)
    - [x] Add tests related to those changes
    - [x] Test py 3.6 to 3.13
    - [X] Ideally don't change current package hash -- **Except 2: armcomputelibrary and mfem**

Commit:
unparser: mirror cpython globals

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython interleave

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython _Precedence enum

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: rename dispatch to traverse

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: rename tree to node

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: change `f: StringIo` to `_source: List[str]`

bench, ubuntu 22.04, python 3.12, unprase all recipes:

before:

  Time (mean ± σ):      4.131 s ±  0.026 s    [User: 4.049 s, System: 0.076 s]
  Range (min … max):    4.074 s …  4.153 s    10 runs

after:
  Time (mean ± σ):      4.130 s ±  0.038 s    [User: 4.041 s, System: 0.082 s]
  Range (min … max):    4.076 s …  4.210 s    10 runs

negligible.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: use contextmanager

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Very minor changes, style, more Python 3 (f-string, nonlocal etc.)

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport "drop redundant parantheses when unparsing tuples as assignment targets"

See https://github.com/python/cpython/commit/51cef8be8c77dff522bec6f95d6853031bf19038

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Partial revert to avoid changing spack package hash

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport "add except*" (python 3.11)

See https://github.com/python/cpython/commit/d60457a6673cf0263213c2f2be02c633ec2e2038

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport _type_params_helper

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport visit_With*

As far as I can tell, since python3.3, With and and WithAsync have items+body,
and withitem has context_expr+optional_vars.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Add comments about cpython difference and py < 3.12 compat

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport NodeVisitor

visit_Interactive and visit_Expression are replaced by generic_visit fallback.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport FunctionType

```
python3 -c 'import ast; print(ast.dump(compile("(int) -> float", "", "func_type", ast.PyCF_ONLY_AST)))'
```

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport "empty sets and NaN"

https://github.com/python/cpython/commit/08ff4369afca84587b1c82034af4e9f64caddbf2

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Backport "lambdas with no parameters"

https://github.com/python/cpython/commit/e4e931a67e49cf3c61263dc94fb0806c34f972cd

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: keyword-only args are available since python 3.0

https://peps.python.org/pep-3102/

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: remove comments

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Partial revert to avoid spack package hash change

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Disabled backport of "round tripping support for typed AST"

Disabled because we don't want to keep type comments for package hashes

https://github.com/python/cpython/commit/dff92bb31f7db1a80ac431811f8108bd0ef9be43

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Disabled backport of "Add docstring support to ast.unparse"

https://github.com/python/cpython/commit/89aa4694fc8c6d190325ef8ed6ce6a6b8efb3e50

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Function return type is avail since py 3.0

https://peps.python.org/pep-3107/

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Fix spack style

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

haampie commented:
Thanks, that is a lot of tedious work.

We should have a GitHub action that computes the package hash for all of builtin and fails when there are inconsistencies.

Commit:
unparser: mirror cpython: Clarify and test 3.11 subscript change

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Remove python 2 support of unicode chars

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: Copy cpython 3.12 JoinedStr visitor as-is

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Patch to support 3.6+ f-strings

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Improve stability of u"" literal strings

The string literal prefix `u""` exists for python 2 compatibility:
https://peps.python.org/pep-0414/

But python <= 3.7 AST does not see `u""`'s kind. So never emit it when
_py_ver_consistent is enabled.

Also remove the workaround for "\u", as it would unparse to different code.
For example `"\ufeff"` would unparse to:
- `u'\ufeff'` in 3.6
- `'\ufeff'` in 3.11

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Fix `3 .__abs__()` in python 3.7

The previous fix did not work because the type `Constant` existed since at least
3.6.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Remove unused future_imports

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Mirror cpython and change package hash

We already changed it in this PR anyway.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Error on format string not compatible with py 3.11

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Mirror cpython 3.12 docstring, fixed for 3.6+

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Mirror cpython 3.8 type_comment, fixed for 3.6+

But wont be executed by spack because requires `compile(... ast.PyCF_TYPE_COMMENTS)`

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Mirror cpython 3.13 type default value

https://github.com/python/cpython/commit/ca269e58c290be8ca11bb728004ea842d9f85e3a

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
package_hash: Fix roundtrip after docstring removal

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Mimic previous unparser to avoid package hash change

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Fix tests with ubuntu's python 3.6

Almalinux 8's python 3.6 has the `kind` attribute.

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython: comprehension is_async avail since python 3.6

On almalinux 8:

```
$ python3 --version
Python 3.6.8
$ python3 -c 'import ast; print(ast.dump(compile("async def f():\n  [a async for a in b]", "", "exec", ast.PyCF_ONLY_AST)))'
Module(body=[AsyncFunctionDef(name='f', args=arguments(args=[], vararg=None, kwonlyargs=[], kw_defaults=[], kwarg=None, defaults=[]), body=[Expr(value=ListComp(elt=Name(id='a', ctx=Load()), generators=[comprehension(target=Name(id='a', ctx=Store()), iter=Name(id='b', ctx=Load()), ifs=[], is_async=1)]))], decorator_list=[], returns=None)])
```

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: mirror cpython 3.14.0a2

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Commit:
unparser: Clarify changes compared to cpython

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

Jordan474 commented:
I believe this pull request is in a good state and ready for review.  See the end of this comment to compare with cpython's ast.py.

The issues with package hash instabilities for Python 3.6 and 3.7 have been resolved.

And I was able to keep previous package hashes (3.8+), **except for `armcomputelibrary` and `mfem`**. This change is due to the resolution of backslashed Unicode characters in the unparse code; for example `'\xef\xbb\xbf'` now roundtrips to `'ï»¿'`. Previously, `spack.util.unparse` manipulated strings to maintain compatibility with Python 2, but I have removed patches for python 2 support, so it now behaves like the built-in `ast.unparse`.

Additionally, all recipes, as well as all of Spack's Python source code (for testing purposes), now roundtrip correctly across all minor Python versions from 3.6 to 3.13. I am uncertain full roundtrip is a desired property (we only need consistent hashes for now), but it works. In other words, as long as the original code is compatible with Python 3.6+, all Python versions should re-emit the same code, that is compatible with python 3.6+.

To reassure myself, I also compared the output of `spack.util.unparse` with `ast.unparse` (Python >= 3.9). Below is a non-exhaustive list of differences between `spack.util.unparse` and `ast.unparse`:

- The `u` string prefix is removed (e.g., `u""` becomes `""`) for Python 3.7 compatibility.
- Lambdas have an extra space added to avoid changes in Spack package hashes (for Python <= 3.10).
- F-string formatted parts switch between quote types to avoid backslashes (`f'{''}'` is only allowed since Python 3.12).
- Unnecessary parentheses around tuples are sometimes added (for Python <= 3.10).

And, I can now much more easily compare spack's unparser directly to `class Unparse` of [cpython Lib/ast.py](https://github.com/python/cpython/blob/main/Lib/ast.py): the difference was reduced from ~1000 line changes to ~200 (includes comments explaining differences). For example:

```console
git clone https://github.com/python/cpython
cp cpython/Lib/ast.py .../spack/lib/spack/spack/util/unparse/cpython-ast.py

cd .../spack/lib/spack/spack/util/unparse
spack style -t black --fix ./cpython-ast.py

your-favorite-diff-tool cpython-ast.py unparser.py  # Skip to the `class Unparse`
```

<details>
<summary>Python versions I tested</summary>

- Python 3.6.8 almalinux 8
- Python 3.6.9 ubuntu 18.04.6
- Python 3.7.17 Spack (ubuntu 24.04)
- Python 3.8.19 Spack (ubuntu 24.04)
- Python 3.9.19 Spack (ubuntu 24.04)
- Python 3.10.14 Spack (ubuntu 24.04)
- Python 3.11.0rc1 ubuntu 24.04
- Python 3.12.3 ubuntu 24.04

</details>

haampie commented:
Will have a look one of these days.

haampie committed:
Fix hash after rebase on develop

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

haampie committed:
Make all Python < 3.8 branches "static"

Signed-off-by: Harmen Stoppels <me@harmenstoppels.nl>

haampie commented:
Meanwhile the `armcomputelibrary` and `mfem` packages were updated, and this PR no longer affects any existing package hash. It's still technically breaking, but not in practice.

haampie commented:
Thanks @Jordan474, sorry it didn't make it into v1.0. I don't think we'll backport it due to the unicode related change, but it will be in v1.1.
